File: python/find_words.py
import sys, copy, signal

def load_dictionnary():
    with open('words.txt', 'r') as dict_file:
        dict = [ word[:-1] for word in dict_file if len(word) > 4 ]
    return dict

def has_letters(word, letters):
    letters = copy.deepcopy(letters)
    for letter in word:
        if not letter in letters:
            return False
        letters.remove(letter)
    return True


def words_with_letters(letters):
    dictionnary = load_dictionnary()
    return [ word for word in dictionnary if has_letters(word, letters) ]


class Engine():
    def __init__(self, letters):
        self.letters = sorted(letters)
        self.found_words = []
        self.possible_words = words_with_letters(letters)

    def is_full(self):
        return len(self.found_words) == len(self.possible_words)

    def guess_word(self, word):
        is_possible = word in self.possible_words
        if is_possible:
            self.found_words.append(word)
        return is_possible

    def pretty_guesses(self):
        words = []
        for w in self.possible_words:
            if w in self.found_words:
                words.append(w)
            else:
                words.append('_' * len(w))
        return '  '.join(words)

    def pretty_solution(self):
        return '  '.join(self.possible_words)

    def pretty_letters(self):
        return ' '.join(self.letters)


if __name__ == '__main__':
    letters = [ 's', 'u', 'm', 'm', 'e', 'r' ]
    engine = Engine(letters)
    print('{}'.format(engine.pretty_guesses()))
    print('{}'.format(engine.pretty_letters()))

    def user_quit(signum, frame):
        print('{}'.format(engine.pretty_solution()))
        exit(1)
    signal.signal(signal.SIGINT, user_quit)

    while not engine.is_full():
        word = sys.stdin.readline()[:-1]
        if engine.guess_word(word):
            print('Correct!\n{}\n{}'.format(
                engine.pretty_guesses(),
                engine.pretty_letters()
            ))
        else:
            print('Wrong!')
    print('You found all the words!')
    exit(0)

File: js/find_words.js

function copy(array){
	var copy = [];
	for(var i=0; i<array.length; i++){
		copy[i] = array[i];
	}
	return copy;
}

function has_letters(word, letters){
	var copied_letters = copy(letters);
	for(var i=0; i<word.length; i++){
		var index = copied_letters.indexOf(word[i]);
		if(index == -1){
			return false;
		}
		copied_letters[index] = null;
	}
	return true;
}

function words_with_letters(dictionnary, letters){
	var words = dictionnary.filter(function(word){
		var test = has_letters(word, letters);
		return test
	});
	return words;
}

function initialize_state(dictionnary, letters){
	var state = {
		letters: letters,
		found_words: [],
		possible_words: words_with_letters(dictionnary, letters)
	};
	return state;
}

function is_full(state){
	return state.found_words.length == state.possible_words.length;
}

function guess_word(state, word){
	if(state.possible_words.includes(word)){
		state.found_words.push(word);
		return true;
	} else {
		return false;
	}
}

function update_display(state){
	var letters_text = state.letters.join(' ');
	var founds = [];
	for(var i=0; i<state.possible_words.length; i++){
		if(state.found_words.includes(state.possible_words[i])){
			founds.push(state.possible_words[i]);
		} else {
			founds.push('_'.repeat(state.possible_words[i].length));
		}
	}
	var founds_text = founds.join(' ');
	$("#display").empty();
	$("#display").append(letters_text);
	$("#display").append("<br>");
	$("#display").append(founds_text);
	$("#display").append("<br>");
};

$.get("./words.txt", function(data) {
	var dictionnary =
		data
		.split('\n')
		.map(function(word){return word.trim();})
		.filter(function(word){return word.length > 2;})
	;
	console.log(dictionnary);
	var state = initialize_state(dictionnary, ['s', 'u', 'm', 'm', 'e', 'r']);
	update_display(state);
	$('#guess-input').on('submit', function(event){
		event.preventDefault();
		var word = $("#guess").val();
		guess_word(state, word);
		update_display(state);
	});
});

